README.md
[![Latest version](https://img.shields.io/crates/v/dmsort.svg)](https://crates.io/crates/dmsort)
[![Documentation](https://docs.rs/dmsort/badge.svg)](https://docs.rs/dmsort)

# Abstract
This is an implementation of a novel [adaptive sorting](https://en.wikipedia.org/wiki/Adaptive_sort) algorithm optimized for nearly-sorted data. Drop-Merge sort is especially useful for when >80% of the data is already in-order, and the unsorted elements are evenly distributed. An example use-case would be re-sorting an already sorted list after minor modifications.

Drop-Merge sort is 2-5 times faster than quicksort when sorting long lists (10k elements or more) in cases where >80% of the data is already in order, while being considerably simpler to implement than other adaptive sorting algorithms.

With `N` elements in the list where `K` elements are out-of-order, Drop-Merge sort performs `O(N + K⋅log(K))` comparisons and use `O(K)` extra memory.

Definition: X% of elements are in order if the [Longest Nondecreasing Subsequence](https://en.wikipedia.org/wiki/Longest_increasing_subsequence) contains X% of the elements.

# Getting started with the crate
Add this to `Cargo.toml`:
```
[dependencies]
dmsort = "1.0.0"
```

And then use it:
``` rust
extern crate dmsort;

fn main() {
	let mut numbers : Vec<i32> = vec!(0, 1, 6, 7, 2, 3, 4, 5);
	dmsort::sort(&mut numbers);
	assert_eq!(numbers, vec!(0, 1, 2, 3, 4, 5, 6, 7));
}
```

[Full dmsort documentation](https://emilk.github.io/drop-merge-sort/dmsort/index.html).

# Performance
To test the performance of this algorithm, I generate almost-sorted data like this (pseudo-code):

```
function generate_test_data(length, disorder_factor) -> Vec {
	result = Vec::new()
	for i in 0..length {
		if random_float() < disorder_factor {
			result.push(random_integer_in_range(0, length))
		} else {
			result.push(i)
		}
	}
	return result
}
```

I then benchmarked Drop-Merge sort against:

* the default Rust sorting algorithm ([Vec::sort](https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#method.sort), a [stable sorting algorithm](https://github.com/rust-lang/rust/pull/38192) based loosely on [Timsort](https://en.wikipedia.org/wiki/Timsort))
* [pdqsort (pattern-defeating quicksort)](https://github.com/stjepang/pdqsort)
* [quickersort (dual-pivot quicksort)](https://github.com/notriddle/quickersort)

The benchmark was run on an MacBook Pro Retina (Early 2013) using `rustc 1.15.0 (10893a9a3 2017-01-19)`.

![Comparing Drop-Merge sort](images/comparisons_1M_i32.png)
![Comparing Drop-Merge sort](images/comparisons_1M_string.png)

We can see that all four algorithms manages to exploit almost-sorted data, but Drop-Merge sort wins out when the disorder factor is less than 30% (more than 70% of the elements are in order).

It also behaves well when the data becomes more random, and at its worst it is still only ~30% slower than pdqsort (which Drop-Merge sort uses as fallback).

Here is another view of the data for 0-50% disorder:

![Speedup over quicksort](images/speedup_1M_i32.png)

Here we can see that we get 5x speedup over quicksort when 99% of the elements are in order, and a 2x speedup when 85% of the elements are in order.

When the disorder is above 30% (less than 70% of the elements are in order), Drop-Merge sort is slower than its competitors.

# Algorithm details
## Background
The paper [Item Retention Improvements to Dropsort, a Lossy Sorting Algorithm](http://micsymposium.org/mics_2011_proceedings/mics2011_submission_13.pdf) by Abram Jackson and Ryan McCulloch introduced improvements to the esoteric "lossy" sorting algorithm known as [*Dropsort*](http://www.dangermouse.net/esoteric/dropsort.html). In Dropsort, out-of-order elements are simply "dropped" (i.e. removed). In the paper, Jackson et al. introduced improvements to Dropsort which improved the detection of out-of-order elements so that more elements would be kept, and fewer dropped.

Although the paper does not spell it out, it does in fact describe a fast `O(N)` algorithm for finding an approximate solution to the [Longest Nondecreasing Subsequence](https://en.wikipedia.org/wiki/Longest_increasing_subsequence) (LNS) problem. The algorithm is very effective when a majority of elements in a list is in order (part of the LNS).

Example input, where most of the data is in order:

![Example of almost-sorted data](images/example.png)

## Drop-Merge sort
The main idea in Drop-Merge sort is this:

* Use the methods described in the Jackson et al. paper to find the Longest Nondecreasing Subsequence (LNS).
* Keep the LNS and drop the outliers into a separate list.
* Sort the list of dropped outliers using a standard sorting algorithm ([`sort_unstable`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable) in this implementation).
* Merge the outliers back into the main list of the already sorted LNS.

Thus despite its heritage, Drop-Merge sort is a *lossless* sorting algorithm (the normal kind).

## Finding the Longest Nondecreasing Subsequence
The implementation uses the idea of *memory* from the Jackson et al. paper to detect *mistakes* - elements that probably should have been dropped instead of accepted. Consider the following sequence of numbers:

`0 1 12 3 4 5 6 7 8 9`

The naïve Dropsort algorithm will accept `0 1 12` and then drop the remaining values (as they are all smaller than 12). The idea of memory is to detect when we have dropped a certain number of elements in a row. When this happens, we roll-back and drop the last element before the long run of drops. This is a form of back-tracking which solves the above problem. In the case of the input above, the algorithm will drop the `12` and keep all other elements. In this implementation, we consider the dropping of 8 elements in a row to be the cutoff point at which we roll-back and undo. This contains the assumption that there will be no more than 8 outliers in a row. It is a very simple method that works well in practice.

Drop-Merge sort will keep in-order elements in situ, moving them towards the start of the list. This helps keep the memory use down and the performance up.

If the elements are not ordered Drop-Merge sort can be around 10% slower than a traditional sorting algorithm. Therefore Drop-Merge sort will try to detect such disordered input and abort, falling back to the standard sorting algorithm.

## Comparison to other adaptive sorting algorithms
An adaptive sorting algorithm is one that can exploit existing order. These algorithms ranges from the complicated to the simple.

On the complicated end there is the famous [Smoothsort](https://en.wikipedia.org/wiki/Smoothsort), which seems, however, to be quite unpopular - probably due to its complexity. I failed to find a good implementation of Smoothsort to compare Drop-Merge sort against. [Timsort](https://en.wikipedia.org/wiki/Timsort) is a more modern and popular adaptive sorting algorithm. It needs long spans of nondecreasing elements to compete with the performance of Drop-Merge sort. The standard Rust sort uses a variant of Timsort, and as you can see from the performance comparisons, Drop-Merge sort wins for the nearly-sorted cases for which it was designed.

On the simple end of the spectrum there are `O(N²)` algorithms that perform extremely well when there are only one or two elements out of place, or the list is very short (a few hundred elements at most). Examples include [Insertion sort](https://en.wikipedia.org/wiki/Insertion_sort) and [Coctail sort](https://en.wikipedia.org/wiki/Cocktail_shaker_sort).

Drop-Merge sort finds an interesting middle-ground – it is reasonably simple (less than 100 lines of code), yet manages to perform well for long lists. Note, however, that Drop-Merge sort depends on another sorting algorithm (e.g. quick-sort) for sorting the out-of-order elements.

## When Drop-Merge sort works best
* Less than 20-30% of the elements out of order AND these are randomly distributed in the data (not clumped).
* You have a lot of data (10k elements or more, and definitively when you get into the millions).


## Limitations and future work
Drop-Merge sort is not stable, which means it will not keep the order of equal elements.

Drop-Merge sort does not sort [in-situ](https://en.wikipedia.org/wiki/In-place_algorithm), but will use `O(K)` extra memory, where `K` is the number of elements out-of-order.

The algorithms uses `recency=8` which means it can handle no more than 8 outliers in a row. This number was chosen by experimentation, and could perhaps be adjusted dynamically for increased performance.

# Other implementations
* C++: [https://github.com/adrian17/cpp-drop-merge-sort](https://github.com/adrian17/cpp-drop-merge-sort)


## TODO
* Set up TravisCI


CHANGELOG.md
# 1.0.2 - 2022-06-19
Fix some bugs found by Miri (https://github.com/emilk/drop-merge-sort/pull/21).

# 1.0.0 - 2018-07-18
Replace dependency on Crate [pdqsort](https://github.com/orlp/pdqsort) with [`sort_unstable`](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable).

# 0.1.3 - 2017-05-08
Update pdqsort to 0.1.2 and update pbr to fix compilation issue due to https://github.com/rust-lang/rust/issues/40952#issuecomment-293720927

# 0.1.2 - 2017-02-05
Update pdqsort to 0.1.1

# 0.1.1 - 2017-01-06
Minor adjustments for faster sorting of shorter lists.

# 0.1.0 - 2017-01-05
Made crate available on cargo.io.


src/main.rs
// By Emil Ernerfeldt December 2016
// LICENSE:
//   This software is dual-licensed to the public domain and under the following
//   license: you are granted a perpetual, irrevocable license to copy, modify,
//   publish, and distribute this file as you see fit.

use std::cell::RefCell;
use std::cmp::Ordering;
use std::fs::File;
use std::io::Write;
use std::panic;
use std::path::Path;
use std::ptr;
use std::collections::BTreeSet;

extern crate rand;
use rand::{Rng, SeedableRng, StdRng};

extern crate pbr;
use pbr::ProgressBar;

extern crate quickersort;
extern crate time;

// ----------------------------------------------------------------------------

/// This speeds up well-ordered input by quite a lot.
const DOUBLE_COMPARISONS : bool = true;

/// Low RECENCY = faster when there is low disorder (a lot of order).
/// High RECENCY = more resilient against long stretches of noise.
/// If RECENCY is too small we are more dependent on nice data/luck.
const RECENCY : usize = 8;

/// Back-track several elements at once. This is helpful when there are big clumps out-of-order.
const FAST_BACKTRACKING : bool = true;

/// Break early if we notice that the input is not ordered enough.
const EARLY_OUT : bool = true;

/// Test for early-out when we have processed len / EARLY_OUT_TEST_AT elements.
const EARLY_OUT_TEST_AT : usize = 4;

/// If more than this percentage of elements have been dropped, we abort.
const EARLY_OUT_DISORDER_FRACTION : f32 = 0.80;

/// Show fastest of BENCH_BEST_OF:
const BENCH_BEST_OF : usize = 5;

static BENCH_RESOLUTION_START  : usize = 20;
static BENCH_RESOLUTION_END    : usize =  99 * 5;
static BENCH_RESOLUTION_CUTOFF : f32   =   0.01;

// ----------------------------------------------------------------------------

/// This is the readable reference implementation that only works for Copy types.
fn dmsort_copy_by<T, F>(slice: &mut [T], mut compare: F) -> usize
	where T: Copy,
		  F: FnMut(&T, &T) -> Ordering
{
	if slice.len() < 2 { return slice.len(); }

	// ------------------------------------------------------------------------
	// First step: heuristically find the Longest Nondecreasing Subsequence (LNS).
	// The LNS is shifted into slice[..write] while slice[write..] will be left unchanged.
	// Elements not part of the LNS will be put in the "dropped" vector.

	let mut dropped = Vec::new();
	let mut num_dropped_in_row = 0;
	let mut write = 0; // Index of where to write the next element to keep.
	let mut read  = 0; // Index of the input stream.

	while read < slice.len() {
		if EARLY_OUT
			&& read == slice.len() / EARLY_OUT_TEST_AT
			&& dropped.len() as f32 > read as f32 * EARLY_OUT_DISORDER_FRACTION {
			// We have seen a lot of the elements and dropped a lot of them.
			// This doesn't look good. Abort.
			for i in 0..dropped.len() {
				slice[write + i] = dropped[i];
			}
			slice.sort_by(|a, b| compare(a, b));
			return dropped.len() * EARLY_OUT_TEST_AT; // Just an estimate.
		}

		if 1 <= write && compare(&slice[read], &slice[write - 1]) == Ordering::Less {
			// The next element is smaller than the last stored one.
			// The question is - should we drop the new element, or was accepting the previous element a mistake?

			/*
			   Check this situation:
			   0 1 2 3 9 5 6 7  (the 9 is a one-off)
			           | |
			           | read
			           write - 1
				Checking this improves performance because we catch common problems earlier (without back-tracking).
			*/
			if DOUBLE_COMPARISONS
				&& num_dropped_in_row == 0
				&& 2 <= write
				&& compare(&slice[read], &slice[write - 2]) != Ordering::Less
			{
				// Quick undo: drop previously accepted element, and overwrite with new one:
				dropped.push(slice[write - 1]);
				slice[write - 1] = slice[read];
				read += 1;
				continue;
			}

			if num_dropped_in_row < RECENCY {
				dropped.push(slice[read]);
				read += 1;
				num_dropped_in_row += 1;
			} else {
				/*
				We accepted something num_dropped_in_row elements back that made us drop all RECENCY subsequent items.
				Accepting that element was obviously a mistake - so let's undo it!

				Example problem (RECENCY = 3):    0 1 12 3 4 5 6
					0 1 12 is accepted. 3, 4, 5 will be rejected because they are larger than the last kept item (12).
					When we get to 5 we reach num_dropped_in_row == RECENCY.
					This will trigger an undo where we drop the 12.
					When we again go to 3, we will keep it because it is larger than the last kept item (1).

				Example worst-case (RECENCY = 3):   ...100 101 102 103 104 1 2 3 4 5 ....
					100-104 is accepted. When we get to 3 we reach num_dropped_in_row == RECENCY.
					We drop 104 and reset the read by RECENCY. We restart, and then we drop again.
					This can lead us to backtracking RECENCY number of elements
					as many times as the leading non-decreasing subsequence is long.
				*/

				// Undo the dropping of elements:
				let trunc_to_length = dropped.len() - num_dropped_in_row;
				dropped.truncate(trunc_to_length);
				read -= num_dropped_in_row;

				let mut num_backtracked = 1;
				write -= 1;

				if FAST_BACKTRACKING { // && 1 <= write && compare(&slice[read], &slice[write - 1]) == Ordering::Less {
					// Back-track until we can accept at least one of the recently dropped elements:
					let max_of_dropped = slice[read..(read + num_dropped_in_row + 1)].iter()
						.max_by(|a, b| compare(a, b)).unwrap();

					while 1 <= write && compare(&max_of_dropped, &slice[write - 1]) == Ordering::Less {
						num_backtracked += 1;
						write -= 1;
					}
				}

				// Drop the back-tracked elements:
				dropped.extend_from_slice(&slice[write..(write + num_backtracked)]);

				num_dropped_in_row = 0;
			}
		} else {
			// Keep:
			slice[write] = slice[read];
			read += 1;
			write += 1;
			num_dropped_in_row = 0;
		}
	}

	let num_dropped = dropped.len();

	// ------------------------------------------------------------------------

	dropped.sort_by(|a, b| compare(a, b));

	// ------------------------------------------------------------------------
	// slice[..write] is now sorted, as is "dropped".
	// We now want to merge these into "slice".
	// Let us do that from the back, putting the largest elements in place first:

	let mut back = slice.len();

	while let Some(&last_dropped) = dropped.last() {
		while 0 < write && compare(&last_dropped, &slice[write - 1]) == Ordering::Less {
			slice[back - 1] = slice[write - 1];
			back -= 1;
			write -= 1;
		}
		slice[back - 1] = last_dropped;
		back -= 1;
		dropped.pop();
	}

	num_dropped
}

fn dmsort_copy<T: Copy + Ord>(slice: &mut [T]) -> usize
{
	dmsort_copy_by(slice, |a, b| a.cmp(b))
}

// ----------------------------------------------------------------------------

/*
A note about protecting us from stack unwinding:

If our compare function panics we need to make sure all objects are put back into slice
so they can be properly destroyed by the caller.

This is done by temporarily bit-copying the data into the dropped vector
and copying them back if there is a panic.
*/
struct DmSorter<'a, T: 'a> {
	/// The slice we are sorting
	slice: &'a mut [T],

	/// Temporary storage of dropped elements.
	dropped: Vec<T>,

	/// Index in self.slice of where to write the next element to keep.
	write: usize,

	// slice[write..(write + dropped.len())] is a gap. The elements can be found in dropped
}

impl<'a, T> Drop for DmSorter<'a, T> {
	fn drop(&mut self) {
		if self.dropped.is_empty() { return; }
		unsafe {
			// This code will only run on stack-unwind (panic).

			// Move back all elements into the slice:
			ptr::copy_nonoverlapping(self.dropped.as_ptr(), &mut self.slice[self.write], self.dropped.len());

			// Make sure the objects aren't destroyed when self.dropped is dropped (avoid-double-free).
			self.dropped.set_len(0);
		}
	}
}

unsafe fn unsafe_push<T>(vec : &mut Vec<T>, value: &T) {
	let old_len = vec.len();
	vec.push(std::mem::uninitialized::<T>());
	ptr::copy_nonoverlapping(value, vec.get_unchecked_mut(old_len), 1);
}

unsafe fn unsafe_copy<T>(slice: &mut [T], source: usize, dest: usize) {
	ptr::copy_nonoverlapping(slice.get_unchecked(source), slice.get_unchecked_mut(dest), 1);
}

fn dmsort_move_by<T, F>(slice: &mut [T], mut compare: F)
	where F: FnMut(&T, &T) -> Ordering
{ unsafe {
	if slice.len() < 2 { return; }

	let mut s = DmSorter{
		slice:   slice,
		dropped: Vec::new(),
		write:   0,
	};

	// ------------------------------------------------------------------------

	let mut num_dropped_in_row = 0;
	let mut read = 0;

	while read < s.slice.len() {
		if EARLY_OUT
			&& read == s.slice.len() / EARLY_OUT_TEST_AT
			&& s.dropped.len() as f32 > read as f32 * EARLY_OUT_DISORDER_FRACTION {
			// We have seen a lot of the elements and dropped a lot of them.
			// This doesn't look good. Abort.
			ptr::copy_nonoverlapping(s.dropped.as_ptr(), &mut s.slice[s.write], s.dropped.len());
			s.dropped.set_len(0);
			s.slice.sort_by(|a, b| compare(a, b));
			return;
		}

		if 1 <= s.write
			&& compare(s.slice.get_unchecked(read), s.slice.get_unchecked(s.write - 1)) == Ordering::Less
		{
			if DOUBLE_COMPARISONS
				&& num_dropped_in_row == 0
				&& 2 <= s.write
				&& compare(s.slice.get_unchecked(read), s.slice.get_unchecked(s.write - 2)) != Ordering::Less
			{
				// Quick undo: drop previously accepted element, and overwrite with new one:
				unsafe_push(&mut s.dropped, s.slice.get_unchecked(s.write - 1));
				unsafe_copy(&mut s.slice, read, s.write - 1);
				read += 1;
				continue;
			}

			if num_dropped_in_row < RECENCY {
				unsafe_push(&mut s.dropped, s.slice.get_unchecked(read));
				read += 1;
				num_dropped_in_row += 1;
			} else {
				// Undo the dropping of elements:
				let trunc_to_length = s.dropped.len() - num_dropped_in_row;
				s.dropped.set_len(trunc_to_length);
				read -= num_dropped_in_row;

				let mut num_backtracked = 1;
				s.write -= 1;

				if FAST_BACKTRACKING {
					// Back-track until we can accept at least one of the recently dropped elements:
					let max_of_dropped = s.slice[read..(read + num_dropped_in_row + 1)].iter()
						.max_by(|a, b| compare(a, b)).unwrap();

					while 1 <= s.write && compare(&max_of_dropped, s.slice.get_unchecked(s.write - 1)) == Ordering::Less {
						num_backtracked += 1;
						s.write -= 1;
					}
				}

				// Append s.slice[read..(read + num_backtracked)] to s.dropped
				{
					let old_len = s.dropped.len();
					for _ in 0..num_backtracked {
						s.dropped.push(std::mem::uninitialized::<T>());
					}
					ptr::copy_nonoverlapping(s.slice.get_unchecked(s.write), s.dropped.get_unchecked_mut(old_len), num_backtracked);
				}

				num_dropped_in_row = 0;
			}
		} else {
			unsafe_copy(&mut s.slice, read, s.write);
			read += 1;
			s.write += 1;
			num_dropped_in_row = 0;
		}
	}

	// ------------------------------------------------------------------------

	s.dropped.sort_by(|a, b| compare(a, b));

	// ------------------------------------------------------------------------
	// s.slice[..s.write] is now sorted, as is "s.dropped".
	// We now want to merge these into "s.slice".
	// Let us do that from the back, putting the largest elements in place first:

	let mut back = s.slice.len();

	loop {
		let old_len = s.dropped.len();
		if old_len == 0 { break; }
		{
			let last_dropped = s.dropped.get_unchecked(old_len - 1);
			while 0 < s.write && compare(last_dropped, s.slice.get_unchecked(s.write - 1)) == Ordering::Less {
				unsafe_copy(&mut s.slice, s.write - 1, back - 1);
				back -= 1;
				s.write -= 1;
			}
			ptr::copy_nonoverlapping(last_dropped, s.slice.get_unchecked_mut(back - 1), 1);
		}
		back -= 1;
		s.dropped.set_len(old_len - 1);
	}
} }

fn dmsort_move<T: Ord>(slice: &mut [T])
{
	dmsort_move_by(slice, |a, b| a.cmp(b))
}

// ----------------------------------------------------------------------------

/// Sorts the elements using the given compare function.
/// Expected number of comparisons is O(N + K * log(K)) where K is the number of elements not in order.
/// Expected memory usage is O(K).
/// Works best for when K < 0.2 * N.
/// The out-of-order elements are expected to be randomly distributed (NOT clumped).
fn dmsort_by<T, F>(slice: &mut [T], compare: F)
	where F: FnMut(&T, &T) -> Ordering
{
	dmsort_move_by(slice, compare);
}

pub fn dmsort_by_key<T, K, F>(slice: &mut [T], key: F)
	where K: Ord,
	      F: Fn(&T) -> K
{
	dmsort_by(slice, |a, b| key(a).cmp(&key(b)));
}

/// Sorts the elements using the given compare function.
/// Expected number of comparisons is O(N + K * log(K)) where K is the number of elements not in order.
/// Expected memory usage is O(K).
/// Works best for when K < 0.2 * N.
/// The out-of-order elements are expected to be randomly distributed (NOT clumped).
fn dmsort<T: Ord>(slice: &mut [T])
{
	dmsort_move_by(slice, |a, b| a.cmp(b));
}

// ----------------------------------------------------------------------------

type Integer = i32;

/// Returns a mostly-sorted array with disorder_factor fraction of elements with random values.
fn generate_integers(rng: &mut rand::StdRng, length: usize, disorder_factor: f32) -> Vec<Integer> {
	let mut result = Vec::with_capacity(length);
	for i in 0..length {
		if rng.next_f32() < disorder_factor {
			result.push(rng.gen_range(0 as Integer, length as Integer));
		} else {
			result.push(i as Integer);
		}
	}
	result
}

fn generate_strings(rng: &mut rand::StdRng, length: usize, disorder_factor: f32) -> Vec<String> {
	generate_integers(rng, length, disorder_factor).iter().map(|&x| format!("{:0100}", x)).collect()
}

fn time_sort_ms<T: Clone, Sorter>(unsorted: &Vec<T>, mut sorter: Sorter) -> (f32, Vec<T>)
	where Sorter: FnMut(&mut Vec<T>)
{
	let mut best_ns = None;
	let mut sorted = Vec::new();

	for _ in 0..BENCH_BEST_OF {
		let mut vec_clone : Vec<T> = unsorted.clone();
		let start_time_ns = time::precise_time_ns();
		sorter(&mut vec_clone);
		let duration_ns = time::precise_time_ns() - start_time_ns;
		if best_ns == None || duration_ns < best_ns.unwrap() {
			best_ns = Some(duration_ns);
		}
		sorted = vec_clone;
	}

	(best_ns.unwrap() as f32 / 1000000.0, sorted)
}

/// Benchmark at these disorders:
fn get_bench_disorders() -> Vec<f32> {
	fn remap(x: usize, in_min: usize, in_max: usize, out_min: f32, out_max: f32) -> f32 {
		out_min + (out_max - out_min) * ((x - in_min) as f32) / ((in_max - in_min) as f32)
	}

	return
		(0..BENCH_RESOLUTION_START).map(    |x| remap(x, 0, BENCH_RESOLUTION_START, 0.0, BENCH_RESOLUTION_CUTOFF)).chain(
		(0..(BENCH_RESOLUTION_END + 1)).map(|x| remap(x, 0, BENCH_RESOLUTION_END,   BENCH_RESOLUTION_CUTOFF, 1.0))
		).collect();
}

fn generate_comparison_data_i32(rng: &mut rand::StdRng, length: usize) {
	let bench_disorders = get_bench_disorders();
	let mut pb = ProgressBar::new(bench_disorders.len() as u64);
	pb.message("Benchmarking integers: ");
	let mut std_file             = File::create(&Path::new("data/i32/std_sort.data")).unwrap();
	let mut quicker_file         = File::create(&Path::new("data/i32/quicker_sort.data")).unwrap();
	let mut dmsort_copy_file     = File::create(&Path::new("data/i32/dmsort_copy_sort.data")).unwrap();
	let mut dmsort_move_file     = File::create(&Path::new("data/i32/dmsort_move_sort.data")).unwrap();
	let mut copy_speedup_file    = File::create(&Path::new("data/i32/dmsort_copy_speedup.data")).unwrap();
	let mut move_speedup_file    = File::create(&Path::new("data/i32/dmsort_move_speedup.data")).unwrap();
	let mut num_dropped_file     = File::create(&Path::new("data/i32/num_dropped.data")).unwrap();

	for disorder_factor in bench_disorders {
		let vec = generate_integers(rng, length, disorder_factor);
		let (std_duration_ms,            std_sorted)            = time_sort_ms(&vec, |x| x.sort());
		let (quicker_duration_ms,        quicker_sorted)        = time_sort_ms(&vec, |x| quickersort::sort(x));
		let (dmsort_copy_duration_ms,    dmsort_copy_sorted)    = time_sort_ms(&vec, |x| {dmsort_copy(x);    ()});
		let (dmsort_move_duration_ms,    dmsort_move_sorted)    = time_sort_ms(&vec, |x| dmsort_move(x));

		let fastest_competitor_ms = std_duration_ms.min(quicker_duration_ms);

		assert_eq!(dmsort_copy_sorted,    std_sorted);
		assert_eq!(dmsort_move_sorted,    std_sorted);
		assert_eq!(quicker_sorted,        std_sorted);

		write!(std_file,             "{} {}\n", disorder_factor * 100.0, std_duration_ms).unwrap();
		write!(quicker_file,         "{} {}\n", disorder_factor * 100.0, quicker_duration_ms).unwrap();
		write!(dmsort_copy_file,     "{} {}\n", disorder_factor * 100.0, dmsort_copy_duration_ms).unwrap();
		write!(dmsort_move_file,     "{} {}\n", disorder_factor * 100.0, dmsort_move_duration_ms).unwrap();
		write!(copy_speedup_file,    "{} {}\n", disorder_factor * 100.0, fastest_competitor_ms / dmsort_copy_duration_ms).unwrap();
		write!(move_speedup_file,    "{} {}\n", disorder_factor * 100.0, fastest_competitor_ms / dmsort_move_duration_ms).unwrap();

		{
			let mut temp = vec.clone();
			let num_dropped_in_row = dmsort_copy(&mut temp);
			let num_dropped = (num_dropped_in_row as f64) / (temp.len() as f64);
			write!(num_dropped_file, "{} {}\n", disorder_factor * 100.0, num_dropped * 100.0).unwrap();
		}

		pb.inc();
	}
}

fn generate_comparison_data_string(rng: &mut rand::StdRng, length: usize) {
	let bench_disorders = get_bench_disorders();
	let mut pb = ProgressBar::new(bench_disorders.len() as u64);
	pb.message("Benchmarking strings: ");
	let mut std_file        = File::create(&Path::new("data/string/std_sort.data")).unwrap();
	let mut quicker_file    = File::create(&Path::new("data/string/quicker_sort.data")).unwrap();
	let mut drop_merge_file = File::create(&Path::new("data/string/drop_merge_sort.data")).unwrap();
	let mut speedup_file    = File::create(&Path::new("data/string/speedup_over_quickersort.data")).unwrap();

	for disorder_factor in bench_disorders {
		let vec = generate_strings(rng, length, disorder_factor);
		let (std_duration_ms,        std_sorted)        = time_sort_ms(&vec, |x| x.sort());
		let (quicker_duration_ms,    quicker_sorted)    = time_sort_ms(&vec, |x| quickersort::sort(x));
		let (drop_merge_duration_ms, drop_merge_sorted) = time_sort_ms(&vec, |x| dmsort(x));

		let fastest_competitor_ms = std_duration_ms.min(quicker_duration_ms);

		assert_eq!(std_sorted, drop_merge_sorted);
		assert_eq!(quicker_sorted, drop_merge_sorted);

		write!(std_file,        "{} {}\n", disorder_factor * 100.0, std_duration_ms).unwrap();
		write!(quicker_file,    "{} {}\n", disorder_factor * 100.0, quicker_duration_ms).unwrap();
		write!(drop_merge_file, "{} {}\n", disorder_factor * 100.0, drop_merge_duration_ms).unwrap();
		write!(speedup_file,    "{} {}\n", disorder_factor * 100.0, fastest_competitor_ms / drop_merge_duration_ms).unwrap();

		pb.inc();
	}
}

#[test]
fn simple_tests() {
	fn test_type<T: Clone + PartialEq + Ord + std::fmt::Debug>(unsorted: Vec<T>) {
		let mut dm_sorted = unsorted.clone();
		dmsort(&mut dm_sorted);

		let mut std_sorted = unsorted.clone();
		std_sorted.sort();

		assert_eq!(dm_sorted, std_sorted, "FAIL with input {:?}", unsorted);
	}

	fn test(list: Vec<Integer>) {
		test_type(list.iter().map(|&x| format!("{:02}", x)).collect());
		test_type(list);
	}

	test(vec!());
	test(vec!(0));
	test(vec!(0, 1));
	test(vec!(1, 0));
	test(vec!(0, 1, 2));
	test(vec!(0, 2, 1));
	test(vec!(1, 0, 2));
	test(vec!(1, 2, 0));
	test(vec!(2, 0, 1));
	test(vec!(2, 1, 0));
	test(vec!(0, 1, 3, 2, 4, -5, 6, 7, 8, 9));
	test(vec!(0, 1, 10, 3, 4, 5, 6, 7, 8, 9));
	test(vec!(10, 1, 2, 3, 4, 5, 6, 7, 8, 9));
	test(vec!(0, 0, 2, 3, 4, 1, 6, 1, 8, 9));
	test(vec!(20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19));
	test(vec!(20, 21, 2, 23, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19));
}

#[test]
fn test_unwind() {
	/*
	The purpose of this test is to make sure that if there is a panic in the compare function
	then we will unwind the stack in such a way that the slice we are sorting
	contains all elements it had when called (but maybe in a different, partially-sorted order).
	This is crucial in order to prevent double-frees.
	*/
	struct TestSortType<'a> {
		id:          usize,
		dropped:     &'a RefCell<BTreeSet<usize>>,
	}
	impl<'a> Drop for TestSortType<'a> {
		fn drop(&mut self) {
			let did_insert = self.dropped.borrow_mut().insert(self.id);
			assert!(did_insert, "Double-free of {}", self.id);
		}
	}

	for break_after_this_many_comparisons in 0..14 {
		let scheuled_panic_code : String = String::from("This is a scheduled panic");

		let dropped = RefCell::new(BTreeSet::new());

		let catch_result = panic::catch_unwind(panic::AssertUnwindSafe(|| {
			let mut data = [
				TestSortType{id: 0, dropped: &dropped},
				TestSortType{id: 1, dropped: &dropped},
				TestSortType{id: 5, dropped: &dropped},
				TestSortType{id: 6, dropped: &dropped},
				TestSortType{id: 2, dropped: &dropped},
				TestSortType{id: 3, dropped: &dropped},
				TestSortType{id: 4, dropped: &dropped},
			];
			let mut num_comparisons = 0;

			dmsort_by(&mut data, |a, b| {
				if num_comparisons == break_after_this_many_comparisons {
					panic!(scheuled_panic_code.clone());
				}
				num_comparisons += 1;
				a.id.cmp(&b.id)
			});
			panic!("We where supposed to abort after {} comparisons, but we only did {}",
			       break_after_this_many_comparisons, num_comparisons);
		}));

		// Make sure we did panic:
		assert!(catch_result.is_err());

		// Make sure we panicked for the right reason:
		let error = catch_result.err().unwrap();
		assert!(error.is::<String>());
		assert_eq!(*error.downcast_ref::<String>().unwrap(), scheuled_panic_code);

		// Make sure we dropped all objects:
		assert_eq!(dropped.borrow_mut().len(), 7);
	}
}

/// Benchmark worst-case input for Drop-Merge sort
fn bench_evil() {
	let evil_input : Vec<_> = (100..1000000).chain(0..100).collect();
	let (std_duration_ms,     std_sorted)     = time_sort_ms(&evil_input, |x| x.sort());
	let (quicker_duration_ms, quicker_sorted) = time_sort_ms(&evil_input, |x| quickersort::sort(x));
	let (drop_duration_ms,    drop_sorted)    = time_sort_ms(&evil_input, |x| dmsort(x));
	// let (drop_duration_ms,    drop_sorted)    = time_sort_ms(&evil_input, |x| {dmsort_copy(x); ()});

	assert_eq!(std_sorted, drop_sorted);
	assert_eq!(std_sorted, quicker_sorted);
	println!("Worst-case input:");
	println!("std::sort:       {} ms", std_duration_ms);
	println!("Quicksort:       {} ms", quicker_duration_ms);
	println!("Drop-Merge sort: {} ms", drop_duration_ms);
}

fn main() {
	bench_evil();

	let seed: &[_] = &[0];
	let mut rng: StdRng = SeedableRng::from_seed(seed);

	generate_comparison_data_i32(&mut rng, 1000000);
	generate_comparison_data_string(&mut rng, 100000);
}
